bokehjs/src/lib/models/annotations/dimensional.ts

```diff
@@ -99,89 +94,130 @@ export abstract class Dimensional extends Model {
     return {
       new_value: preferred_value,
       new_unit,
-      new_long_unit,
       scale_factor,
       exact,
     }
   }
 }
 
+export namespace CustomDimensional {
+  export type Attrs = p.AttrsOf<Props>
+  export type Props = Dimensional.Props & {
+    basis: p.Property<DictLike<[number, string]>>
+  }
+}
+
+export interface CustomDimensional extends CustomDimensional.Attrs {}
+
+export abstract class CustomDimensional extends Dimensional {
+  declare properties: CustomDimensional.Props
+
+  constructor(attrs?: Partial<CustomDimensional.Attrs>) {
+    super(attrs)
+  }
+
+  static {
+    this.define<CustomDimensional.Props>(({Dict, Tuple, Number, String}) => ({
+      basis: [ Dict(Tuple(Number, String)) ],
+    }))
+  }
+
+  get_basis(): DictLike<[number, string]> {
+    return this.basis
+  }
+}
+
 export namespace Metric {
   export type Attrs = p.AttrsOf<Props>
-  export type Props = Dimensional.Props
+  export type Props = Dimensional.Props & {
+    base_unit: p.Property<string>
+  }
 }
 
 export interface Metric extends Metric.Attrs {}
 
-export abstract class Metric extends Dimensional {
+export class Metric extends Dimensional {
   declare properties: Metric.Props
 
   constructor(attrs?: Partial<Metric.Attrs>) {
     super(attrs)
   }
 
   static {
+    this.define<Metric.Props>(({String}) => ({
+      base_unit: [ String ],
+    }))
+
     this.override<Metric.Props>({
       ticks: [1, 2, 5, 10, 15, 20, 25, 50, 75, 100, 125, 150, 200, 250, 500, 750],
     })
   }
 
-  build_basis(): Basis {
-    const {_basis_template, _short_name, _long_name, _tex_repr} = this
-    const {include, exclude} = this
+  static _metric_basis = [
+    ["Q", 1e30,  "Q"],
+    ["R", 1e27,  "R"],
+    ["Y", 1e24,  "Y"],
+    ["Z", 1e21,  "Z"],
+    ["E", 1e18,  "E"],
+    ["P", 1e15,  "P"],
+    ["T", 1e12,  "T"],
+    ["G", 1e9,   "G"],
+    ["M", 1e6,   "M"],
+    ["k", 1e3,   "k"],
+    ["h", 1e2,   "h"],
+    ["",  1e0,   ""],
+    ["d", 1e-1,  "d"],
+    ["c", 1e-2,  "c"],
+    ["m", 1e-3,  "m"],
+    ["µ", 1e-6,  "\\mu"],
+    ["n", 1e-9,  "n"],
+    ["p", 1e-12, "p"],
+    ["f", 1e-15, "f"],
+    ["a", 1e-18, "a"],
+    ["z", 1e-21, "z"],
+    ["y", 1e-24, "y"],
+    ["r", 1e-27, "r"],
+    ["q", 1e-30, "q"],
+  ] as const
```

Comment:
If we are not going to store the "long" names in the structure for any use any longer, it would be nice to add comments like `// quetta` after each case

Reply:
I restored long names, at least in the API. How to use them is a concern for the future.